File: backend/product-client/src/main/java/org/andante/speakers/dto/SpeakersVariantOutputDTO.java
package org.andante.speakers.dto;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import lombok.experimental.SuperBuilder;
import org.andante.product.dto.ProductVariantOutputDTO;

@SuperBuilder
@Data
@AllArgsConstructor
@NoArgsConstructor
@EqualsAndHashCode(callSuper = true)
public class SpeakersVariantOutputDTO extends ProductVariantOutputDTO {

    private Integer loudness;
    private String color;
    private Long speakersId;
}

File: frontend/src/store/api/result/dto/product/variant/SpeakersVariantOutputDTO.ts
import { ProductVariantOutputDTO } from "./ProductVariantOutputDTO";

export interface SpeakersVariantOutputDTO extends ProductVariantOutputDTO {
    loudness: number,
    color: string,
    speakersId: number
};

export function isSpeakersVariantOutputDTO(variant: ProductVariantOutputDTO): variant is SpeakersVariantOutputDTO {
    return "speakersId" in variant;
}

export function isSpeakersVariantOutputDTOArray(variants: ProductVariantOutputDTO[]): variants is SpeakersVariantOutputDTO[] {
    return variants.every(isSpeakersVariantOutputDTO);
}